// @carbonplan/icechunk-js/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  CompressionAlgorithm: () => CompressionAlgorithm,
  FileType: () => FileType,
  HeaderParseError: () => HeaderParseError,
  HttpStorage: () => HttpStorage,
  IcechunkStore: () => IcechunkStore,
  NotFoundError: () => NotFoundError,
  ReadSession: () => ReadSession,
  Repository: () => Repository,
  SpecVersion: () => SpecVersion,
  StorageError: () => StorageError,
  decodeBase32: () => decodeBase32,
  decodeObjectId12: () => decodeObjectId12,
  decodeObjectId8: () => decodeObjectId8,
  encodeBase32: () => encodeBase32,
  encodeObjectId12: () => encodeObjectId12,
  encodeObjectId8: () => encodeObjectId8
});
module.exports = __toCommonJS(index_exports);

// src/storage/storage.ts
var NotFoundError = class extends Error {
  constructor(path) {
    super(`Object not found: ${path}`);
    this.path = path;
    this.name = "NotFoundError";
  }
};
var StorageError = class extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "StorageError";
  }
};
var AbortError = class extends Error {
  constructor(message = "Operation aborted") {
    super(message);
    this.name = "AbortError";
  }
};

// src/format/constants.ts
var PATHS = {
  /** Refs directory */
  REFS: "refs",
  /** Snapshots directory */
  SNAPSHOTS: "snapshots",
  /** Manifests directory */
  MANIFESTS: "manifests",
  /** Chunks directory */
  CHUNKS: "chunks"
};
var REPO_INFO_PATH = "repo";
function getBranchRefDirPath(name) {
  return `${PATHS.REFS}/branch.${name}/`;
}
function getTagRefDirPath(name) {
  return `${PATHS.REFS}/tag.${name}/`;
}
var REF_FILE_NAME = "ref.json";
function getBranchRefPath(name) {
  return `${PATHS.REFS}/branch.${name}/${REF_FILE_NAME}`;
}
function getTagRefPath(name) {
  return `${PATHS.REFS}/tag.${name}/${REF_FILE_NAME}`;
}
function getSnapshotPath(id) {
  return `${PATHS.SNAPSHOTS}/${id}`;
}
function getManifestPath(id) {
  return `${PATHS.MANIFESTS}/${id}`;
}
function getChunkPath(id) {
  return `${PATHS.CHUNKS}/${id}`;
}

// src/format/object-id.ts
var ENCODE_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
var DECODE_MAP = {};
for (let i = 0; i < ENCODE_ALPHABET.length; i++) {
  const char = ENCODE_ALPHABET[i];
  DECODE_MAP[char] = i;
  DECODE_MAP[char.toLowerCase()] = i;
}
DECODE_MAP["I"] = DECODE_MAP["i"] = 1;
DECODE_MAP["L"] = DECODE_MAP["l"] = 1;
DECODE_MAP["O"] = DECODE_MAP["o"] = 0;
function encodeBase32(bytes) {
  if (bytes.length === 0) return "";
  let result = "";
  let buffer = 0;
  let bitsLeft = 0;
  for (const byte of bytes) {
    buffer = buffer << 8 | byte;
    bitsLeft += 8;
    while (bitsLeft >= 5) {
      bitsLeft -= 5;
      const index = buffer >> bitsLeft & 31;
      result += ENCODE_ALPHABET[index];
    }
  }
  if (bitsLeft > 0) {
    const index = buffer << 5 - bitsLeft & 31;
    result += ENCODE_ALPHABET[index];
  }
  return result;
}
function decodeBase32(str) {
  if (str.length === 0) return new Uint8Array(0);
  str = str.replace(/-/g, "");
  const result = [];
  let buffer = 0;
  let bitsLeft = 0;
  for (const char of str) {
    const value = DECODE_MAP[char];
    if (value === void 0) {
      throw new Error(`Invalid Base32 Crockford character: ${char}`);
    }
    buffer = buffer << 5 | value;
    bitsLeft += 5;
    if (bitsLeft >= 8) {
      bitsLeft -= 8;
      result.push(buffer >> bitsLeft & 255);
    }
  }
  return new Uint8Array(result);
}
function encodeObjectId12(id) {
  if (id.length !== 12) {
    throw new Error(`Expected 12 bytes, got ${id.length}`);
  }
  return encodeBase32(id);
}
function decodeObjectId12(str) {
  const bytes = decodeBase32(str);
  if (bytes.length !== 12) {
    throw new Error(`Expected 12 bytes after decoding, got ${bytes.length}`);
  }
  return bytes;
}
function encodeObjectId8(id) {
  if (id.length !== 8) {
    throw new Error(`Expected 8 bytes, got ${id.length}`);
  }
  return encodeBase32(id);
}
function decodeObjectId8(str) {
  const bytes = decodeBase32(str);
  if (bytes.length !== 8) {
    throw new Error(`Expected 8 bytes after decoding, got ${bytes.length}`);
  }
  return bytes;
}

// src/format/flatbuffers/repo-parser.ts
var import_flatbuffers = require("flatbuffers");

// src/format/flatbuffers/generated/metadata-item.ts
var MetadataItem = class _MetadataItem {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsMetadataItem(bb, obj) {
    return (obj || new _MetadataItem()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  name(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  value(index) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
  }
  valueLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  valueArray() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? new Uint8Array(
      this.bb.bytes().buffer,
      this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset),
      this.bb.__vector_len(this.bb_pos + offset)
    ) : null;
  }
};

// src/format/flatbuffers/generated/ref.ts
var Ref = class _Ref {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsRef(bb, obj) {
    return (obj || new _Ref()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  name(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  snapshotIndex() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/repo-status.ts
var RepoStatus = class _RepoStatus {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsRepoStatus(bb, obj) {
    return (obj || new _RepoStatus()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  availability() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint8(this.bb_pos + offset) : 0 /* Online */;
  }
  setAt() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  limitedAvailabilityReason(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
};

// src/format/flatbuffers/generated/object-id12.ts
var ObjectId12 = class {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  bytes(index) {
    return this.bb.readUint8(this.bb_pos + 0 + index);
  }
  static sizeOf() {
    return 12;
  }
};

// src/format/flatbuffers/generated/snapshot-info.ts
var SnapshotInfo = class _SnapshotInfo {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsSnapshotInfo(bb, obj) {
    return (obj || new _SnapshotInfo()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  id(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  parentOffset() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readInt32(this.bb_pos + offset) : 0;
  }
  flushedAt() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  message(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  metadata(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? (obj || new MetadataItem()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  metadataLength() {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/update.ts
var Update = class _Update {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsUpdate(bb, obj) {
    return (obj || new _Update()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  updateTypeType() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint8(this.bb_pos + offset) : 0 /* NONE */;
  }
  updateType(obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__union(obj, this.bb_pos + offset) : null;
  }
  updatedAt() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  backupPath(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
};

// src/format/flatbuffers/generated/repo.ts
var Repo = class _Repo {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsRepo(bb, obj) {
    return (obj || new _Repo()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  specVersion() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
  }
  tags(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new Ref()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  tagsLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  branches(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? (obj || new Ref()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  branchesLength() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  deletedTags(index, optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.__string(
      this.bb.__vector(this.bb_pos + offset) + index * 4,
      optionalEncoding
    ) : null;
  }
  deletedTagsLength() {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  snapshots(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? (obj || new SnapshotInfo()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  snapshotsLength() {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  status(obj) {
    const offset = this.bb.__offset(this.bb_pos, 14);
    return offset ? (obj || new RepoStatus()).__init(
      this.bb.__indirect(this.bb_pos + offset),
      this.bb
    ) : null;
  }
  metadata(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 16);
    return offset ? (obj || new MetadataItem()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  metadataLength() {
    const offset = this.bb.__offset(this.bb_pos, 16);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  latestUpdates(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 18);
    return offset ? (obj || new Update()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  latestUpdatesLength() {
    const offset = this.bb.__offset(this.bb_pos, 18);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  repoBeforeUpdates(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 20);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
};

// src/format/flatbuffers/repo-parser.ts
var import_fzstd = require("fzstd");

// src/format/header.ts
var HEADER_SIZE = 39;
var MAGIC_BYTES = new Uint8Array([
  73,
  67,
  69,
  // ICE
  240,
  159,
  167,
  138,
  // 🧊 (U+1F9CA in UTF-8)
  67,
  72,
  85,
  78,
  75
  // CHUNK
]);
var SpecVersion = /* @__PURE__ */ ((SpecVersion3) => {
  SpecVersion3[SpecVersion3["V1_0"] = 1] = "V1_0";
  SpecVersion3[SpecVersion3["V2_0"] = 2] = "V2_0";
  return SpecVersion3;
})(SpecVersion || {});
var FileType = /* @__PURE__ */ ((FileType2) => {
  FileType2[FileType2["Snapshot"] = 1] = "Snapshot";
  FileType2[FileType2["Manifest"] = 2] = "Manifest";
  FileType2[FileType2["Attributes"] = 3] = "Attributes";
  FileType2[FileType2["TransactionLog"] = 4] = "TransactionLog";
  FileType2[FileType2["Chunk"] = 5] = "Chunk";
  FileType2[FileType2["RepoInfo"] = 6] = "RepoInfo";
  return FileType2;
})(FileType || {});
var CompressionAlgorithm = /* @__PURE__ */ ((CompressionAlgorithm2) => {
  CompressionAlgorithm2[CompressionAlgorithm2["None"] = 0] = "None";
  CompressionAlgorithm2[CompressionAlgorithm2["Zstd"] = 1] = "Zstd";
  return CompressionAlgorithm2;
})(CompressionAlgorithm || {});
var HeaderParseError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "HeaderParseError";
  }
};
function parseHeader(data) {
  if (data.length < HEADER_SIZE) {
    throw new HeaderParseError(
      `Buffer too small: expected at least ${HEADER_SIZE} bytes, got ${data.length}`
    );
  }
  const magic = data.slice(0, 12);
  if (!compareMagic(magic, MAGIC_BYTES)) {
    throw new HeaderParseError("Invalid magic bytes: not an icechunk file");
  }
  const implementationBytes = data.slice(12, 36);
  const implementation = new TextDecoder().decode(implementationBytes).trimEnd();
  const specVersionByte = data[36];
  if (specVersionByte !== 1 /* V1_0 */ && specVersionByte !== 2 /* V2_0 */) {
    throw new HeaderParseError(`Invalid spec version: ${specVersionByte}`);
  }
  const specVersion = specVersionByte;
  const fileTypeByte = data[37];
  if (fileTypeByte < 1 || fileTypeByte > 6) {
    throw new HeaderParseError(`Invalid file type: ${fileTypeByte}`);
  }
  const fileType = fileTypeByte;
  const compressionByte = data[38];
  if (compressionByte !== 0 /* None */ && compressionByte !== 1 /* Zstd */) {
    throw new HeaderParseError(
      `Invalid compression algorithm: ${compressionByte}`
    );
  }
  const compression = compressionByte;
  return {
    implementation,
    specVersion,
    fileType,
    compression
  };
}
function validateFileType(header, expectedType) {
  if (header.fileType !== expectedType) {
    const expected = FileType[expectedType];
    const actual = FileType[header.fileType];
    throw new HeaderParseError(
      `Invalid file type: expected ${expected}, got ${actual}`
    );
  }
}
function getDataAfterHeader(data) {
  return data.slice(HEADER_SIZE);
}
function compareMagic(a, b) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

// src/format/flatbuffers/repo-parser.ts
var SUPPORTED_SPEC_VERSION = 2;
function parseRepo(data) {
  if (data.length < HEADER_SIZE) {
    throw new Error(
      `Repo file too small: ${data.length} bytes, need at least ${HEADER_SIZE}`
    );
  }
  const header = parseHeader(data);
  validateFileType(header, 6 /* RepoInfo */);
  let fbData = getDataAfterHeader(data);
  if (header.compression === 1 /* Zstd */) {
    fbData = (0, import_fzstd.decompress)(fbData);
  }
  const bb = new import_flatbuffers.ByteBuffer(fbData);
  const repo = Repo.getRootAsRepo(bb);
  const specVersion = repo.specVersion();
  if (specVersion !== SUPPORTED_SPEC_VERSION) {
    throw new Error(
      `Unsupported repo spec version: ${specVersion}, expected ${SUPPORTED_SPEC_VERSION}`
    );
  }
  const snapshotsLength = repo.snapshotsLength();
  return { repo, specVersion, snapshotsLength };
}
function compareUtf8ByteOrder(aBytes, bBytes) {
  const minLen = Math.min(aBytes.length, bBytes.length);
  for (let i = 0; i < minLen; i++) {
    if (aBytes[i] < bBytes[i]) return -1;
    if (aBytes[i] > bBytes[i]) return 1;
  }
  return aBytes.length - bBytes.length;
}
function binarySearchRef(parsedRepo, accessor, length, name) {
  const { snapshotsLength } = parsedRepo;
  if (length === 0) return null;
  const targetBytes = new TextEncoder().encode(name);
  let low = 0;
  let high = length - 1;
  while (low <= high) {
    const mid = low + high >>> 1;
    const refTable = accessor(mid);
    if (!refTable) {
      throw new Error(`Corrupted repo file: null ref table at index ${mid}`);
    }
    const refName = refTable.name();
    if (refName === null) {
      throw new Error(`Corrupted repo file: null ref name at index ${mid}`);
    }
    const refNameBytes = new TextEncoder().encode(refName);
    const cmp = compareUtf8ByteOrder(refNameBytes, targetBytes);
    if (cmp < 0) {
      low = mid + 1;
    } else if (cmp > 0) {
      high = mid - 1;
    } else {
      const snapshotIndex = refTable.snapshotIndex();
      if (snapshotIndex >= snapshotsLength) {
        throw new Error(
          `Invalid snapshot index ${snapshotIndex} for ref '${name}', snapshots array has ${snapshotsLength} entries`
        );
      }
      return getSnapshotIdByIndex(parsedRepo, snapshotIndex);
    }
  }
  return null;
}
function getSnapshotIdByIndex(parsedRepo, index) {
  const snapshotInfo = parsedRepo.repo.snapshots(index);
  if (!snapshotInfo) {
    throw new Error(`Corrupted repo file: null snapshot at index ${index}`);
  }
  const idObj = snapshotInfo.id();
  if (!idObj) {
    throw new Error(`Corrupted repo file: snapshot ${index} missing id`);
  }
  return idObj.bb.bytes().slice(idObj.bb_pos, idObj.bb_pos + 12);
}
function listRefs(accessor, length) {
  const names = [];
  for (let i = 0; i < length; i++) {
    const refTable = accessor(i);
    if (!refTable) {
      throw new Error(`Corrupted repo file: null ref table at index ${i}`);
    }
    const name = refTable.name();
    if (name === null) {
      throw new Error(`Corrupted repo file: null ref name at index ${i}`);
    }
    names.push(name);
  }
  return names;
}
function resolveBranch(parsedRepo, name) {
  const { repo } = parsedRepo;
  return binarySearchRef(
    parsedRepo,
    (i) => repo.branches(i),
    repo.branchesLength(),
    name
  );
}
function resolveTag(parsedRepo, name) {
  const { repo } = parsedRepo;
  return binarySearchRef(
    parsedRepo,
    (i) => repo.tags(i),
    repo.tagsLength(),
    name
  );
}
function listBranchesFromRepo(parsedRepo) {
  const { repo } = parsedRepo;
  return listRefs((i) => repo.branches(i), repo.branchesLength());
}
function listTagsFromRepo(parsedRepo) {
  const { repo } = parsedRepo;
  return listRefs((i) => repo.tags(i), repo.tagsLength());
}

// src/reader/session.ts
var import_fzstd2 = require("fzstd");

// src/format/flatbuffers/types.ts
function asObjectId12(bytes) {
  if (bytes.length !== 12) {
    throw new Error(
      `Invalid ObjectId12: expected 12 bytes, got ${bytes.length}`
    );
  }
  return bytes;
}
function asObjectId8(bytes) {
  if (bytes.length !== 8) {
    throw new Error(`Invalid ObjectId8: expected 8 bytes, got ${bytes.length}`);
  }
  return bytes;
}

// src/format/flatbuffers/snapshot-parser.ts
var import_flatbuffers2 = require("flatbuffers");

// src/format/flatbuffers/generated/manifest-file-info.ts
var ManifestFileInfo = class {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  id(obj) {
    return (obj || new ObjectId12()).__init(this.bb_pos, this.bb);
  }
  sizeBytes() {
    return this.bb.readUint64(this.bb_pos + 16);
  }
  numChunkRefs() {
    return this.bb.readUint32(this.bb_pos + 24);
  }
  static sizeOf() {
    return 32;
  }
};

// src/format/flatbuffers/generated/dimension-name.ts
var DimensionName = class _DimensionName {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsDimensionName(bb, obj) {
    return (obj || new _DimensionName()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  name(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
};

// src/format/flatbuffers/generated/dimension-shape.ts
var DimensionShape = class {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  arrayLength() {
    return this.bb.readUint64(this.bb_pos);
  }
  chunkLength() {
    return this.bb.readUint64(this.bb_pos + 8);
  }
  static sizeOf() {
    return 16;
  }
};

// src/format/flatbuffers/generated/chunk-index-range.ts
var ChunkIndexRange = class {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  from() {
    return this.bb.readUint32(this.bb_pos);
  }
  to() {
    return this.bb.readUint32(this.bb_pos + 4);
  }
  static sizeOf() {
    return 8;
  }
};

// src/format/flatbuffers/generated/manifest-ref.ts
var ManifestRef = class _ManifestRef {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsManifestRef(bb, obj) {
    return (obj || new _ManifestRef()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  objectId(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  extents(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new ChunkIndexRange()).__init(
      this.bb.__vector(this.bb_pos + offset) + index * 8,
      this.bb
    ) : null;
  }
  extentsLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/array-node-data.ts
var ArrayNodeData = class _ArrayNodeData {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsArrayNodeData(bb, obj) {
    return (obj || new _ArrayNodeData()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  shape(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new DimensionShape()).__init(
      this.bb.__vector(this.bb_pos + offset) + index * 16,
      this.bb
    ) : null;
  }
  shapeLength() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  dimensionNames(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new DimensionName()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  dimensionNamesLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  manifests(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? (obj || new ManifestRef()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  manifestsLength() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/object-id8.ts
var ObjectId8 = class {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  bytes(index) {
    return this.bb.readUint8(this.bb_pos + 0 + index);
  }
  static sizeOf() {
    return 8;
  }
};

// src/format/flatbuffers/generated/node-snapshot.ts
var NodeSnapshot = class _NodeSnapshot {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsNodeSnapshot(bb, obj) {
    return (obj || new _NodeSnapshot()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  id(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId8()).__init(this.bb_pos + offset, this.bb) : null;
  }
  path(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  userData(index) {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
  }
  userDataLength() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  userDataArray() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? new Uint8Array(
      this.bb.bytes().buffer,
      this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset),
      this.bb.__vector_len(this.bb_pos + offset)
    ) : null;
  }
  nodeDataType() {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.readUint8(this.bb_pos + offset) : 0 /* NONE */;
  }
  nodeData(obj) {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? this.bb.__union(obj, this.bb_pos + offset) : null;
  }
};

// src/format/flatbuffers/generated/snapshot.ts
var Snapshot = class _Snapshot {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsSnapshot(bb, obj) {
    return (obj || new _Snapshot()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  id(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  parentId(obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  nodes(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? (obj || new NodeSnapshot()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  nodesLength() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  flushedAt() {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  message(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  metadata(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 14);
    return offset ? (obj || new MetadataItem()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  metadataLength() {
    const offset = this.bb.__offset(this.bb_pos, 14);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  manifestFiles(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 16);
    return offset ? (obj || new ManifestFileInfo()).__init(
      this.bb.__vector(this.bb_pos + offset) + index * 32,
      this.bb
    ) : null;
  }
  manifestFilesLength() {
    const offset = this.bb.__offset(this.bb_pos, 16);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/snapshot-parser.ts
function readId12(bb, bbPos) {
  return bb.bytes().slice(bbPos, bbPos + 12);
}
function readId8(bb, bbPos) {
  return bb.bytes().slice(bbPos, bbPos + 8);
}
function parseSnapshot(data) {
  const bb = new import_flatbuffers2.ByteBuffer(data);
  const fbsSnapshot = Snapshot.getRootAsSnapshot(bb);
  const idObj = fbsSnapshot.id();
  if (!idObj) throw new Error("Snapshot missing required id field");
  const id = asObjectId12(readId12(idObj.bb, idObj.bb_pos));
  const parentIdObj = fbsSnapshot.parentId();
  const parentId = parentIdObj ? asObjectId12(readId12(parentIdObj.bb, parentIdObj.bb_pos)) : null;
  const nodesLength = fbsSnapshot.nodesLength();
  const nodes = [];
  for (let i = 0; i < nodesLength; i++) {
    const fbsNode = fbsSnapshot.nodes(i);
    if (fbsNode) {
      nodes.push(parseNodeSnapshot(fbsNode));
    }
  }
  const flushedAt = fbsSnapshot.flushedAt();
  const message = fbsSnapshot.message() ?? "";
  const metadataLength = fbsSnapshot.metadataLength();
  const metadata = [];
  for (let i = 0; i < metadataLength; i++) {
    const fbsMeta = fbsSnapshot.metadata(i);
    if (fbsMeta) {
      const name = fbsMeta.name();
      const value = fbsMeta.valueArray();
      if (name && value) {
        metadata.push({ name, value: new Uint8Array(value) });
      }
    }
  }
  const manifestFilesLength = fbsSnapshot.manifestFilesLength();
  const manifestFiles = [];
  for (let i = 0; i < manifestFilesLength; i++) {
    const fbsMfi = fbsSnapshot.manifestFiles(i);
    if (fbsMfi) {
      const mfiIdObj = fbsMfi.id();
      if (!mfiIdObj) throw new Error("ManifestFileInfo missing id");
      manifestFiles.push({
        id: asObjectId12(readId12(mfiIdObj.bb, mfiIdObj.bb_pos)),
        sizeBytes: Number(fbsMfi.sizeBytes()),
        numChunkRefs: fbsMfi.numChunkRefs()
      });
    }
  }
  return {
    id,
    parentId,
    nodes,
    flushedAt,
    message,
    metadata,
    manifestFiles
  };
}
function parseNodeSnapshot(fbsNode) {
  const idObj = fbsNode.id();
  if (!idObj) throw new Error("NodeSnapshot missing required id field");
  const id = asObjectId8(readId8(idObj.bb, idObj.bb_pos));
  const path = fbsNode.path() ?? "";
  const userData = fbsNode.userDataArray() ?? new Uint8Array(0);
  const nodeDataType = fbsNode.nodeDataType();
  const nodeData = parseNodeData(fbsNode, nodeDataType);
  return { id, path, userData, nodeData };
}
function parseNodeData(fbsNode, unionType) {
  if (unionType === 0 /* NONE */ || unionType === 2 /* Group */) {
    return { type: "group" };
  }
  if (unionType === 1 /* Array */) {
    const arrayData = fbsNode.nodeData(new ArrayNodeData());
    if (!arrayData) {
      throw new Error("ArrayNodeData union type but no table");
    }
    return parseArrayNodeData(arrayData);
  }
  throw new Error(`Unknown node data union type: ${unionType}`);
}
function parseArrayNodeData(fbsArray) {
  const shapeLength = fbsArray.shapeLength();
  const shape = [];
  for (let i = 0; i < shapeLength; i++) {
    const fbsShape = fbsArray.shape(i);
    if (fbsShape) {
      shape.push({
        arrayLength: Number(fbsShape.arrayLength()),
        chunkLength: Number(fbsShape.chunkLength())
      });
    }
  }
  const dimNamesLength = fbsArray.dimensionNamesLength();
  const dimensionNames = [];
  for (let i = 0; i < dimNamesLength; i++) {
    const fbsDimName = fbsArray.dimensionNames(i);
    if (fbsDimName) {
      dimensionNames.push(fbsDimName.name());
    } else {
      dimensionNames.push(null);
    }
  }
  const manifestsLength = fbsArray.manifestsLength();
  const manifests = [];
  for (let i = 0; i < manifestsLength; i++) {
    const fbsManifest = fbsArray.manifests(i);
    if (fbsManifest) {
      manifests.push(parseManifestRef(fbsManifest));
    }
  }
  return {
    type: "array",
    shape,
    dimensionNames,
    manifests
  };
}
function parseManifestRef(fbsRef) {
  const objectIdObj = fbsRef.objectId();
  if (!objectIdObj) throw new Error("ManifestRef missing object_id");
  const objectId = asObjectId12(readId12(objectIdObj.bb, objectIdObj.bb_pos));
  const extentsLength = fbsRef.extentsLength();
  const extents = [];
  for (let i = 0; i < extentsLength; i++) {
    const fbsExtent = fbsRef.extents(i);
    if (fbsExtent) {
      extents.push({ from: fbsExtent.from(), to: fbsExtent.to() });
    }
  }
  return { objectId, extents };
}

// src/format/flatbuffers/manifest-parser.ts
var import_flatbuffers3 = require("flatbuffers");

// src/format/flatbuffers/generated/chunk-ref.ts
var ChunkRef = class _ChunkRef {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsChunkRef(bb, obj) {
    return (obj || new _ChunkRef()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  index(index) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.readUint32(this.bb.__vector(this.bb_pos + offset) + index * 4) : 0;
  }
  indexLength() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  indexArray() {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? new Uint32Array(
      this.bb.bytes().buffer,
      this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset),
      this.bb.__vector_len(this.bb_pos + offset)
    ) : null;
  }
  inline(index) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
  }
  inlineLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
  inlineArray() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? new Uint8Array(
      this.bb.bytes().buffer,
      this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset),
      this.bb.__vector_len(this.bb_pos + offset)
    ) : null;
  }
  offset() {
    const offset = this.bb.__offset(this.bb_pos, 8);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  length() {
    const offset = this.bb.__offset(this.bb_pos, 10);
    return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt("0");
  }
  chunkId(obj) {
    const offset = this.bb.__offset(this.bb_pos, 12);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  location(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 14);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  checksumEtag(optionalEncoding) {
    const offset = this.bb.__offset(this.bb_pos, 16);
    return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
  }
  checksumLastModified() {
    const offset = this.bb.__offset(this.bb_pos, 18);
    return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/array-manifest.ts
var ArrayManifest = class _ArrayManifest {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsArrayManifest(bb, obj) {
    return (obj || new _ArrayManifest()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  nodeId(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId8()).__init(this.bb_pos + offset, this.bb) : null;
  }
  refs(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new ChunkRef()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  refsLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/generated/manifest.ts
var Manifest = class _Manifest {
  bb = null;
  bb_pos = 0;
  __init(i, bb) {
    this.bb_pos = i;
    this.bb = bb;
    return this;
  }
  static getRootAsManifest(bb, obj) {
    return (obj || new _Manifest()).__init(
      bb.readInt32(bb.position()) + bb.position(),
      bb
    );
  }
  id(obj) {
    const offset = this.bb.__offset(this.bb_pos, 4);
    return offset ? (obj || new ObjectId12()).__init(this.bb_pos + offset, this.bb) : null;
  }
  arrays(index, obj) {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? (obj || new ArrayManifest()).__init(
      this.bb.__indirect(
        this.bb.__vector(this.bb_pos + offset) + index * 4
      ),
      this.bb
    ) : null;
  }
  arraysLength() {
    const offset = this.bb.__offset(this.bb_pos, 6);
    return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
  }
};

// src/format/flatbuffers/manifest-parser.ts
function parseManifest(data) {
  const bb = new import_flatbuffers3.ByteBuffer(data);
  const fbsManifest = Manifest.getRootAsManifest(bb);
  const idObj = fbsManifest.id();
  if (!idObj) throw new Error("Manifest missing required id field");
  const id = asObjectId12(
    idObj.bb.bytes().slice(idObj.bb_pos, idObj.bb_pos + 12)
  );
  const arraysLength = fbsManifest.arraysLength();
  const arrays = [];
  for (let i = 0; i < arraysLength; i++) {
    const fbsArray = fbsManifest.arrays(i);
    if (fbsArray) {
      arrays.push(parseArrayManifest(fbsArray));
    }
  }
  return { id, arrays };
}
function parseArrayManifest(fbsArray) {
  const nodeIdObj = fbsArray.nodeId();
  if (!nodeIdObj)
    throw new Error("ArrayManifest missing required node_id field");
  const nodeId = asObjectId8(
    nodeIdObj.bb.bytes().slice(nodeIdObj.bb_pos, nodeIdObj.bb_pos + 8)
  );
  const refsLength = fbsArray.refsLength();
  const refs = [];
  for (let i = 0; i < refsLength; i++) {
    const fbsRef = fbsArray.refs(i);
    if (fbsRef) {
      refs.push(parseChunkRef(fbsRef));
    }
  }
  return { nodeId, refs };
}
function parseChunkRef(fbsRef) {
  const indexLength = fbsRef.indexLength();
  const index = [];
  for (let i = 0; i < indexLength; i++) {
    index.push(fbsRef.index(i));
  }
  const inlineData = fbsRef.inlineArray();
  const inline = inlineData ? new Uint8Array(inlineData) : null;
  const offset = Number(fbsRef.offset());
  const length = Number(fbsRef.length());
  const chunkIdObj = fbsRef.chunkId();
  const chunkId = chunkIdObj ? asObjectId12(
    chunkIdObj.bb.bytes().slice(chunkIdObj.bb_pos, chunkIdObj.bb_pos + 12)
  ) : null;
  const location = fbsRef.location();
  const checksumEtag = fbsRef.checksumEtag();
  const checksumLastModified = fbsRef.checksumLastModified();
  return {
    index,
    inline,
    offset,
    length,
    chunkId,
    location,
    checksumEtag,
    checksumLastModified
  };
}
function findChunkRef(manifest, nodeId, coords) {
  const arrayManifest = binarySearchArrayManifest(manifest.arrays, nodeId);
  if (!arrayManifest) return null;
  return binarySearchChunkRef(arrayManifest.refs, coords);
}
function binarySearchArrayManifest(arrays, nodeId) {
  let low = 0;
  let high = arrays.length - 1;
  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const cmp = compareBytes(arrays[mid].nodeId, nodeId);
    if (cmp === 0) {
      return arrays[mid];
    } else if (cmp < 0) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return null;
}
function compareBytes(a, b) {
  const minLen = Math.min(a.length, b.length);
  for (let i = 0; i < minLen; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}
function compareCoords(a, b) {
  const minLen = Math.min(a.length, b.length);
  for (let i = 0; i < minLen; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}
function binarySearchChunkRef(refs, coords) {
  let low = 0;
  let high = refs.length - 1;
  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const cmp = compareCoords(refs[mid].index, coords);
    if (cmp === 0) {
      return refs[mid];
    } else if (cmp < 0) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return null;
}
function getChunkPayload(ref) {
  if (ref.inline !== null) {
    return { type: "inline", data: ref.inline };
  }
  if (ref.chunkId !== null) {
    return {
      type: "native",
      chunkId: ref.chunkId,
      offset: ref.offset,
      length: ref.length
    };
  }
  if (ref.location !== null) {
    return {
      type: "virtual",
      location: ref.location,
      offset: ref.offset,
      length: ref.length,
      checksumEtag: ref.checksumEtag,
      checksumLastModified: ref.checksumLastModified
    };
  }
  throw new Error("Invalid ChunkRef: no inline, chunkId, or location");
}

// src/format/flatbuffers/metadata.ts
var import_msgpack = require("@msgpack/msgpack");
var flexbuffers = __toESM(require("flatbuffers/js/flexbuffers.js"), 1);
function deserializeMetadata(items, specVersion) {
  const result = /* @__PURE__ */ Object.create(null);
  for (const item of items) {
    result[item.name] = specVersion === 1 /* V1_0 */ ? (0, import_msgpack.decode)(item.value) : flexbuffers.toObject(
      item.value.buffer.slice(
        item.value.byteOffset,
        item.value.byteOffset + item.value.byteLength
      )
    );
  }
  return result;
}

// src/reader/session.ts
var ReadSession = class _ReadSession {
  storage;
  snapshot;
  specVersion;
  manifestCache = /* @__PURE__ */ new Map();
  constructor(storage, snapshot, specVersion) {
    this.storage = storage;
    this.snapshot = snapshot;
    this.specVersion = specVersion;
  }
  /**
   * Open a read session for a specific snapshot.
   *
   * @param storage - Storage backend
   * @param snapshotId - Snapshot ID (12 bytes)
   * @param options - Optional request options (signal for cancellation)
   * @returns ReadSession instance
   */
  static async open(storage, snapshotId, options) {
    const { snapshot, specVersion } = await _ReadSession.loadSnapshot(
      storage,
      snapshotId,
      options
    );
    return new _ReadSession(storage, snapshot, specVersion);
  }
  /** Load and parse a snapshot from storage */
  static async loadSnapshot(storage, snapshotId, options) {
    const path = getSnapshotPath(encodeObjectId12(snapshotId));
    const data = await storage.getObject(path, void 0, options);
    const header = parseHeader(data);
    validateFileType(header, 1 /* Snapshot */);
    let flatbufferData = getDataAfterHeader(data);
    if (header.compression === 1 /* Zstd */) {
      flatbufferData = (0, import_fzstd2.decompress)(flatbufferData);
    }
    return {
      snapshot: parseSnapshot(flatbufferData),
      specVersion: header.specVersion
    };
  }
  /** Load and parse a manifest from storage */
  async loadManifest(manifestId, options) {
    const idStr = encodeObjectId12(manifestId);
    const cached = this.manifestCache.get(idStr);
    if (cached) return cached;
    const path = getManifestPath(idStr);
    const data = await this.storage.getObject(path, void 0, options);
    const header = parseHeader(data);
    validateFileType(header, 2 /* Manifest */);
    let flatbufferData = getDataAfterHeader(data);
    if (header.compression === 1 /* Zstd */) {
      flatbufferData = (0, import_fzstd2.decompress)(flatbufferData);
    }
    const manifest = parseManifest(flatbufferData);
    this.manifestCache.set(idStr, manifest);
    return manifest;
  }
  /**
   * Get the snapshot ID.
   */
  getSnapshotId() {
    return this.snapshot.id;
  }
  /**
   * Get the spec version of the snapshot.
   */
  getSpecVersion() {
    return this.specVersion;
  }
  /**
   * Get the commit message for this snapshot.
   */
  getMessage() {
    return this.snapshot.message;
  }
  /**
   * Get the timestamp when this snapshot was created.
   */
  getFlushedAt() {
    return new Date(Number(this.snapshot.flushedAt / 1000n));
  }
  /**
   * Get deserialized snapshot metadata.
   *
   * Decodes MessagePack (v1) or FlexBuffers (v2) metadata items
   * into a plain key-value object.
   */
  getSnapshotMetadata() {
    return deserializeMetadata(this.snapshot.metadata, this.specVersion);
  }
  /**
   * Get a node by path.
   *
   * @param path - Absolute path (e.g., "/array" or "/group/nested")
   * @returns NodeSnapshot or null if not found
   */
  getNode(path) {
    const normalizedPath = path.startsWith("/") ? path : `/${path}`;
    return this.binarySearchNode(normalizedPath);
  }
  /**
   * List all nodes in the snapshot.
   *
   * @returns Array of all nodes
   */
  listNodes() {
    return [...this.snapshot.nodes];
  }
  /**
   * List children of a group.
   *
   * @param parentPath - Path to the parent group (use "/" for root)
   * @returns Array of child nodes
   */
  listChildren(parentPath) {
    const normalizedParent = parentPath === "/" ? "" : parentPath;
    const prefix = normalizedParent + "/";
    return this.snapshot.nodes.filter((node) => {
      if (!node.path.startsWith(prefix)) return false;
      const rest = node.path.slice(prefix.length);
      return !rest.includes("/");
    });
  }
  /**
   * Get the Zarr metadata for a node.
   *
   * @param path - Path to the node
   * @returns Parsed JSON metadata or null if node not found
   */
  getMetadata(path) {
    const node = this.getNode(path);
    if (!node) return null;
    const json = new TextDecoder().decode(node.userData);
    return JSON.parse(json);
  }
  /**
   * Get raw user data (Zarr metadata bytes) for a node.
   *
   * @param path - Path to the node
   * @returns Raw metadata bytes or null if node not found
   */
  getRawMetadata(path) {
    const node = this.getNode(path);
    if (!node) return null;
    return node.userData;
  }
  /**
   * Read chunk data for an array.
   *
   * @param path - Path to the array
   * @param coords - Chunk coordinates (N-dimensional)
   * @param options - Optional request options (signal for cancellation)
   * @returns Chunk data bytes or null if not found
   */
  async getChunk(path, coords, options) {
    if (options?.signal?.aborted) return null;
    const node = this.getNode(path);
    if (!node || node.nodeData.type !== "array") {
      return null;
    }
    const arrayData = node.nodeData;
    try {
      for (const manifestRef of arrayData.manifests) {
        if (!this.coordsInExtents(coords, manifestRef.extents)) {
          continue;
        }
        const manifest = await this.loadManifest(manifestRef.objectId, options);
        const chunkRef = findChunkRef(manifest, node.id, coords);
        if (!chunkRef) continue;
        const payload = getChunkPayload(chunkRef);
        return this.fetchChunkPayload(payload, options);
      }
    } catch (error) {
      if (error instanceof AbortError) return null;
      throw error;
    }
    return null;
  }
  /**
   * Read a byte range of chunk data for an array.
   *
   * Like getChunk, but fetches only the requested byte range from storage
   * instead of the full chunk. Used by IcechunkStore.getRange to support
   * zarrita's sharded array reads efficiently.
   *
   * @param path - Path to the array
   * @param coords - Chunk coordinates (N-dimensional)
   * @param range - Byte range within the chunk data
   * @param options - Optional request options (signal for cancellation)
   * @returns Chunk data bytes or null if not found
   */
  async getChunkRange(path, coords, range, options) {
    if (options?.signal?.aborted) return null;
    const node = this.getNode(path);
    if (!node || node.nodeData.type !== "array") {
      return null;
    }
    const arrayData = node.nodeData;
    try {
      for (const manifestRef of arrayData.manifests) {
        if (!this.coordsInExtents(coords, manifestRef.extents)) {
          continue;
        }
        const manifest = await this.loadManifest(manifestRef.objectId, options);
        const chunkRef = findChunkRef(manifest, node.id, coords);
        if (!chunkRef) continue;
        const payload = getChunkPayload(chunkRef);
        return this.fetchChunkPayloadRange(payload, range, options);
      }
    } catch (error) {
      if (error instanceof AbortError) return null;
      throw error;
    }
    return null;
  }
  /** Check if coordinates fall within extent ranges */
  coordsInExtents(coords, extents) {
    if (coords.length !== extents.length) return false;
    for (let i = 0; i < coords.length; i++) {
      const { from, to } = extents[i];
      if (coords[i] < from || coords[i] >= to) {
        return false;
      }
    }
    return true;
  }
  /** Fetch chunk data based on payload type */
  async fetchChunkPayload(payload, options) {
    switch (payload.type) {
      case "inline":
        return payload.data;
      case "native": {
        const path = getChunkPath(encodeObjectId12(payload.chunkId));
        const range = {
          start: payload.offset,
          end: payload.offset + payload.length
        };
        const data = await this.storage.getObject(path, range, options);
        if (data.length === payload.length) return data;
        if (data.length >= range.end) {
          return data.slice(range.start, range.end);
        }
        throw new Error(
          `Storage returned ${data.length} bytes for ${path} range ${range.start}-${range.end - 1}; expected ${payload.length} bytes`
        );
      }
      case "virtual": {
        let httpUrl = translateToHttpUrl(payload.location);
        let fetchInit = {
          headers: {
            Range: `bytes=${payload.offset}-${payload.offset + payload.length - 1}`
          },
          signal: options?.signal
        };
        if (options?.transformRequest) {
          const result = await options.transformRequest(httpUrl, {
            method: "GET"
          });
          httpUrl = result.url;
          if (result.headers) {
            fetchInit.headers = { ...fetchInit.headers, ...result.headers };
          }
          const { url: _, headers: __, method: ___, ...rest } = result;
          fetchInit = { ...fetchInit, ...rest };
        }
        let response;
        try {
          response = await fetch(httpUrl, fetchInit);
        } catch (error) {
          if (error instanceof Error && error.name === "AbortError") {
            throw new AbortError();
          }
          throw error;
        }
        if (response.status !== 200 && response.status !== 206) {
          throw new Error(
            `Failed to fetch virtual chunk from ${httpUrl}: ${response.status} ${response.statusText}`
          );
        }
        const data = new Uint8Array(await response.arrayBuffer());
        if (response.status === 206) {
          if (data.length !== payload.length) {
            throw new Error(
              `Virtual range response size mismatch for ${httpUrl}: expected ${payload.length} bytes, got ${data.length}`
            );
          }
          return data;
        }
        const absoluteEnd = payload.offset + payload.length;
        if (data.length >= absoluteEnd) {
          return data.slice(payload.offset, absoluteEnd);
        }
        throw new Error(
          `Virtual range request not honored for ${httpUrl}: need at least ${absoluteEnd} bytes for fallback slicing, got ${data.length}`
        );
      }
    }
  }
  /** Fetch a byte range of chunk data based on payload type */
  async fetchChunkPayloadRange(payload, range, options) {
    let rangeStart;
    let rangeEnd;
    if ("suffixLength" in range) {
      rangeStart = payload.type === "inline" ? payload.data.length - range.suffixLength : payload.length - range.suffixLength;
      rangeEnd = payload.type === "inline" ? payload.data.length : payload.length;
    } else {
      rangeStart = range.offset;
      rangeEnd = range.offset + range.length;
    }
    switch (payload.type) {
      case "inline":
        return payload.data.slice(rangeStart, rangeEnd);
      case "native": {
        const path = getChunkPath(encodeObjectId12(payload.chunkId));
        const storageRange = {
          start: payload.offset + rangeStart,
          end: payload.offset + rangeEnd
        };
        const expectedSize = rangeEnd - rangeStart;
        const data = await this.storage.getObject(path, storageRange, options);
        if (data.length === expectedSize) return data;
        if (data.length >= storageRange.end) {
          return data.slice(storageRange.start, storageRange.end);
        }
        throw new Error(
          `Storage returned ${data.length} bytes for ${path} range ${storageRange.start}-${storageRange.end - 1}; expected ${expectedSize} bytes`
        );
      }
      case "virtual": {
        const absoluteStart = payload.offset + rangeStart;
        const absoluteEnd = payload.offset + rangeEnd;
        let httpUrl = translateToHttpUrl(payload.location);
        let fetchInit = {
          headers: {
            Range: `bytes=${absoluteStart}-${absoluteEnd - 1}`
          },
          signal: options?.signal
        };
        if (options?.transformRequest) {
          const result = await options.transformRequest(httpUrl, {
            method: "GET"
          });
          httpUrl = result.url;
          if (result.headers) {
            fetchInit.headers = { ...fetchInit.headers, ...result.headers };
          }
          const { url: _, headers: __, method: ___, ...rest } = result;
          fetchInit = { ...fetchInit, ...rest };
        }
        let response;
        try {
          response = await fetch(httpUrl, fetchInit);
        } catch (error) {
          if (error instanceof Error && error.name === "AbortError") {
            throw new AbortError();
          }
          throw error;
        }
        if (response.status !== 200 && response.status !== 206) {
          throw new Error(
            `Failed to fetch virtual chunk from ${httpUrl}: ${response.status} ${response.statusText}`
          );
        }
        const expectedSize = rangeEnd - rangeStart;
        const data = new Uint8Array(await response.arrayBuffer());
        if (response.status === 206) {
          if (data.length !== expectedSize) {
            throw new Error(
              `Virtual range response size mismatch for ${httpUrl}: expected ${expectedSize} bytes, got ${data.length}`
            );
          }
          return data;
        }
        if (data.length >= absoluteEnd) {
          return data.slice(absoluteStart, absoluteEnd);
        }
        throw new Error(
          `Virtual range request not honored for ${httpUrl}: need at least ${absoluteEnd} bytes for fallback slicing, got ${data.length}`
        );
      }
    }
  }
  /** Binary search for a node by path */
  binarySearchNode(path) {
    const nodes = this.snapshot.nodes;
    let low = 0;
    let high = nodes.length - 1;
    while (low <= high) {
      const mid = Math.floor((low + high) / 2);
      const cmp = compareUtf8Bytes(nodes[mid].path, path);
      if (cmp === 0) {
        return nodes[mid];
      } else if (cmp < 0) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }
    return null;
  }
};
var utf8Encoder = new TextEncoder();
function compareUtf8Bytes(a, b) {
  const bytesA = utf8Encoder.encode(a);
  const bytesB = utf8Encoder.encode(b);
  const minLen = Math.min(bytesA.length, bytesB.length);
  for (let i = 0; i < minLen; i++) {
    if (bytesA[i] !== bytesB[i]) {
      return bytesA[i] - bytesB[i];
    }
  }
  return bytesA.length - bytesB.length;
}
function translateToHttpUrl(url) {
  if (url.startsWith("s3://")) {
    const rest = url.slice(5);
    const slashIndex = rest.indexOf("/");
    if (slashIndex === -1) {
      const bucket2 = rest;
      if (bucket2.includes(".")) {
        return `https://s3.amazonaws.com/${bucket2}/`;
      }
      return `https://${bucket2}.s3.amazonaws.com/`;
    }
    const bucket = rest.slice(0, slashIndex);
    const key = rest.slice(slashIndex + 1);
    if (bucket.includes(".")) {
      return `https://s3.amazonaws.com/${bucket}/${key}`;
    }
    return `https://${bucket}.s3.amazonaws.com/${key}`;
  }
  if (url.startsWith("gs://") || url.startsWith("gcs://")) {
    const prefixLen = url.startsWith("gs://") ? 5 : 6;
    const rest = url.slice(prefixLen);
    return `https://storage.googleapis.com/${rest}`;
  }
  return url;
}

// src/reader/repository.ts
var Repository = class _Repository {
  storage;
  repoInfo = null;
  repoInfoAttempted = false;
  constructor(storage) {
    this.storage = storage;
  }
  /**
   * Load and cache the v2 repo info file.
   *
   * Uses getObject() directly to avoid race conditions with exists().
   * - NotFoundError => v1 format (no repo file)
   * - Any other error => hard error (parse failure, etc.)
   *
   * @param options - Optional request options (signal for cancellation)
   * @returns ParsedRepo if v2 format, null if v1 format
   * @throws Error if repo file exists but fails to parse
   * @throws AbortError if the operation was aborted
   */
  async loadRepoInfo(options) {
    if (this.repoInfoAttempted) {
      return this.repoInfo;
    }
    this.repoInfoAttempted = true;
    try {
      const data = await this.storage.getObject(
        REPO_INFO_PATH,
        void 0,
        options
      );
      this.repoInfo = parseRepo(data);
      return this.repoInfo;
    } catch (error) {
      if (error instanceof AbortError) {
        this.repoInfoAttempted = false;
        throw error;
      }
      if (error instanceof NotFoundError) {
        return null;
      }
      throw new Error(
        `Failed to parse v2 repo file: ${error instanceof Error ? error.message : error}`
      );
    }
  }
  /**
   * Open an icechunk repository.
   *
   * - Try to load and parse repo info file (v2+ format)
   * - If NotFoundError, check for main branch (v1 format)
   * - Parse failures are hard errors (matching Rust behavior)
   *
   * @param options - Repository options including storage backend
   * @param requestOptions - Optional request options (signal for cancellation)
   * @returns A Repository instance
   */
  static async open(options, requestOptions) {
    const repo = new _Repository(options.storage);
    if (options.formatVersion === "v1") {
      repo.repoInfoAttempted = true;
      return repo;
    }
    const repoInfo = await repo.loadRepoInfo(requestOptions);
    if (repoInfo) {
      return repo;
    }
    if (options.formatVersion === "v2") {
      throw new Error("Repository info not found but v2 format was specified");
    }
    const mainBranchDir = getBranchRefDirPath("main");
    const mainExists = await repo.hasAnyRefFile(
      mainBranchDir,
      getBranchRefPath("main"),
      requestOptions
    );
    if (mainExists) {
      return repo;
    }
    throw new Error(
      "Not a valid icechunk repository: neither repo info file nor main branch found"
    );
  }
  /**
   * Check if any ref file exists in a directory.
   * Ref files are .json files that are not .deleted files.
   *
   * @param dirPrefix - Directory prefix to check
   * @param legacyPath - Optional legacy ref.json path to check if listing fails
   * @param options - Optional request options (signal for cancellation)
   */
  async hasAnyRefFile(dirPrefix, legacyPath, options) {
    try {
      for await (const path of this.storage.listPrefix(dirPrefix)) {
        if (path.endsWith(".json") && !path.endsWith(".deleted")) {
          return true;
        }
      }
    } catch (error) {
      if (error instanceof AbortError) {
        throw error;
      }
      if (legacyPath) {
        return await this.storage.exists(legacyPath, options);
      }
    }
    return false;
  }
  /**
   * Find the latest non-deleted ref file in a directory.
   * When storage supports listPrefix(), refs may use versioned filenames
   * (e.g., AAAAAAAA.json) for optimistic concurrency control. The latest
   * version has the highest filename. Deletion tombstones are checked.
   *
   * When listing is not supported (e.g., HTTP storage), only the legacy
   * ref.json path is checked (no tombstone check - assumes read-only).
   *
   * @param dirPrefix - Directory prefix to search
   * @param legacyPath - Optional legacy ref.json path to try if listing fails
   */
  async findLatestRefFile(dirPrefix, legacyPath) {
    const jsonFiles = [];
    const deletedFiles = /* @__PURE__ */ new Set();
    try {
      for await (const path of this.storage.listPrefix(dirPrefix)) {
        if (path.endsWith(".deleted")) {
          deletedFiles.add(path.slice(0, -".deleted".length));
        } else if (path.endsWith(".json")) {
          jsonFiles.push(path);
        }
      }
    } catch (error) {
      if (error instanceof AbortError) {
        throw error;
      }
      return legacyPath || null;
    }
    if (jsonFiles.length === 0) {
      return null;
    }
    jsonFiles.sort();
    for (let i = jsonFiles.length - 1; i >= 0; i--) {
      if (!deletedFiles.has(jsonFiles[i])) {
        return jsonFiles[i];
      }
    }
    return null;
  }
  /**
   * List all branches in the repository.
   *
   * Branches with deletion tombstones on their latest ref file are excluded.
   *
   * @returns Array of branch names
   */
  async listBranches() {
    const repoInfo = await this.loadRepoInfo();
    if (repoInfo) {
      return listBranchesFromRepo(repoInfo);
    }
    const branchFiles = /* @__PURE__ */ new Map();
    try {
      for await (const path of this.storage.listPrefix(
        `${PATHS.REFS}/branch.`
      )) {
        const jsonMatch = path.match(/^refs\/branch\.(.+)\/([^/]+\.json)$/);
        const deletedMatch = path.match(
          /^refs\/branch\.(.+)\/([^/]+\.json)\.deleted$/
        );
        if (deletedMatch) {
          const branchName = deletedMatch[1];
          const refFile = `refs/branch.${branchName}/${deletedMatch[2]}`;
          if (!branchFiles.has(branchName)) {
            branchFiles.set(branchName, {
              jsonFiles: [],
              deletedFiles: /* @__PURE__ */ new Set()
            });
          }
          branchFiles.get(branchName).deletedFiles.add(refFile);
        } else if (jsonMatch) {
          const branchName = jsonMatch[1];
          if (!branchFiles.has(branchName)) {
            branchFiles.set(branchName, {
              jsonFiles: [],
              deletedFiles: /* @__PURE__ */ new Set()
            });
          }
          branchFiles.get(branchName).jsonFiles.push(path);
        }
      }
    } catch {
      throw new Error("Cannot list branches: storage does not support listing");
    }
    const result = [];
    for (const [branchName, { jsonFiles, deletedFiles }] of branchFiles) {
      if (jsonFiles.length === 0) continue;
      jsonFiles.sort();
      const latestRef = jsonFiles[jsonFiles.length - 1];
      if (!deletedFiles.has(latestRef)) {
        result.push(branchName);
      }
    }
    return result;
  }
  /**
   * List all tags in the repository.
   *
   * Tags with deletion tombstones on their latest ref file are excluded.
   *
   * @returns Array of tag names
   */
  async listTags() {
    const repoInfo = await this.loadRepoInfo();
    if (repoInfo) {
      return listTagsFromRepo(repoInfo);
    }
    const tagFiles = /* @__PURE__ */ new Map();
    try {
      for await (const path of this.storage.listPrefix(`${PATHS.REFS}/tag.`)) {
        const jsonMatch = path.match(/^refs\/tag\.(.+)\/([^/]+\.json)$/);
        const deletedMatch = path.match(
          /^refs\/tag\.(.+)\/([^/]+\.json)\.deleted$/
        );
        if (deletedMatch) {
          const tagName = deletedMatch[1];
          const refFile = `refs/tag.${tagName}/${deletedMatch[2]}`;
          if (!tagFiles.has(tagName)) {
            tagFiles.set(tagName, { jsonFiles: [], deletedFiles: /* @__PURE__ */ new Set() });
          }
          tagFiles.get(tagName).deletedFiles.add(refFile);
        } else if (jsonMatch) {
          const tagName = jsonMatch[1];
          if (!tagFiles.has(tagName)) {
            tagFiles.set(tagName, { jsonFiles: [], deletedFiles: /* @__PURE__ */ new Set() });
          }
          tagFiles.get(tagName).jsonFiles.push(path);
        }
      }
    } catch {
      throw new Error("Cannot list tags: storage does not support listing");
    }
    const result = [];
    for (const [tagName, { jsonFiles, deletedFiles }] of tagFiles) {
      if (jsonFiles.length === 0) continue;
      jsonFiles.sort();
      const latestRef = jsonFiles[jsonFiles.length - 1];
      if (!deletedFiles.has(latestRef)) {
        result.push(tagName);
      }
    }
    return result;
  }
  /**
   * Checkout a branch to get a read session.
   *
   * @param name - Branch name
   * @param options - Optional request options (signal for cancellation)
   * @returns Read session at the branch's current snapshot
   */
  async checkoutBranch(name, options) {
    const repoInfo = await this.loadRepoInfo(options);
    if (repoInfo) {
      const snapshotId2 = resolveBranch(repoInfo, name);
      if (!snapshotId2) {
        throw new Error(`Branch not found: ${name}`);
      }
      return ReadSession.open(this.storage, snapshotId2, options);
    }
    const refDirPath = getBranchRefDirPath(name);
    const refPath = await this.findLatestRefFile(
      refDirPath,
      getBranchRefPath(name)
    );
    if (!refPath) {
      throw new Error(`Reference not found: ${refDirPath}`);
    }
    const snapshotId = await this.readSnapshotIdFromRef(refPath, options);
    return ReadSession.open(this.storage, snapshotId, options);
  }
  /**
   * Checkout a tag to get a read session.
   *
   * @param name - Tag name
   * @param options - Optional request options (signal for cancellation)
   * @returns Read session at the tag's snapshot
   */
  async checkoutTag(name, options) {
    const repoInfo = await this.loadRepoInfo(options);
    if (repoInfo) {
      const snapshotId2 = resolveTag(repoInfo, name);
      if (!snapshotId2) {
        throw new Error(`Tag not found: ${name}`);
      }
      return ReadSession.open(this.storage, snapshotId2, options);
    }
    const refDirPath = getTagRefDirPath(name);
    const legacyPath = getTagRefPath(name);
    const refPath = await this.findLatestRefFile(refDirPath, legacyPath);
    if (!refPath) {
      throw new Error(`Reference not found: ${refDirPath}`);
    }
    if (refPath === legacyPath && await this.storage.exists(`${refPath}.deleted`, options)) {
      throw new Error(`Tag not found: ${name}`);
    }
    const snapshotId = await this.readSnapshotIdFromRef(refPath, options);
    return ReadSession.open(this.storage, snapshotId, options);
  }
  /**
   * Checkout a specific snapshot by ID.
   *
   * @param snapshotId - Snapshot ID (12 bytes or Base32 string)
   * @param options - Optional request options (signal for cancellation)
   * @returns Read session at the specified snapshot
   */
  async checkoutSnapshot(snapshotId, options) {
    const id = typeof snapshotId === "string" ? decodeObjectId12(snapshotId) : snapshotId;
    return ReadSession.open(this.storage, id, options);
  }
  /**
   * Get the storage backend.
   */
  getStorage() {
    return this.storage;
  }
  /** Read and parse a ref file */
  async readRef(path, options) {
    const data = await this.storage.getObject(path, void 0, options);
    const json = new TextDecoder().decode(data);
    return JSON.parse(json);
  }
  /** Read snapshot ID from a ref file */
  async readSnapshotIdFromRef(path, options) {
    try {
      const ref = await this.readRef(path, options);
      return decodeObjectId12(ref.snapshot);
    } catch (error) {
      if (error instanceof NotFoundError) {
        throw new Error(`Reference not found: ${path}`);
      }
      throw error;
    }
  }
};

// src/storage/http-storage.ts
var HttpStorage = class {
  baseUrl;
  options;
  /**
   * Create an HTTP storage backend.
   *
   * @param baseUrl - Base URL for the repository (e.g., "https://example.com/repo")
   * @param options - Additional options
   */
  constructor(baseUrl, options = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.options = options;
  }
  /** Build full URL for a path */
  getUrl(path) {
    const normalizedPath = path.startsWith("/") ? path.slice(1) : path;
    return `${this.baseUrl}/${normalizedPath}`;
  }
  /** Build headers for a request */
  getHeaders(range) {
    const headers = { ...this.options.headers };
    if (range) {
      headers["Range"] = `bytes=${range.start}-${range.end - 1}`;
    }
    return headers;
  }
  async getObject(path, range, options) {
    if (options?.signal?.aborted) {
      throw new AbortError();
    }
    const url = this.getUrl(path);
    const headers = this.getHeaders(range);
    let response;
    try {
      response = await fetch(url, {
        method: "GET",
        headers,
        credentials: this.options.credentials,
        cache: this.options.cache,
        signal: options?.signal
      });
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw new AbortError();
      }
      throw new StorageError(
        `Failed to fetch ${url}: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : void 0
      );
    }
    if (response.status === 404) {
      throw new NotFoundError(path);
    }
    if (response.status !== 200 && response.status !== 206) {
      throw new StorageError(
        `HTTP ${response.status} ${response.statusText} for ${url}`
      );
    }
    const arrayBuffer = await response.arrayBuffer();
    return new Uint8Array(arrayBuffer);
  }
  async exists(path, options) {
    if (options?.signal?.aborted) {
      throw new AbortError();
    }
    const url = this.getUrl(path);
    try {
      const response = await fetch(url, {
        method: "HEAD",
        headers: this.options.headers,
        credentials: this.options.credentials,
        signal: options?.signal
      });
      return response.ok;
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw new AbortError();
      }
      return false;
    }
  }
  async *listPrefix(_prefix) {
    throw new StorageError(
      "Listing not supported for HTTP storage. Use S3Storage for listing."
    );
  }
};

// src/store.ts
var IcechunkStore = class _IcechunkStore {
  session;
  transformRequest;
  constructor(session, transformRequest) {
    if (!(session instanceof ReadSession)) {
      throw new Error(
        "IcechunkStore constructor is private. Use IcechunkStore.open() instead."
      );
    }
    this.session = session;
    this.transformRequest = transformRequest;
  }
  static async open(arg, options = {}) {
    if (arg instanceof ReadSession) {
      return new _IcechunkStore(arg, options.transformRequest);
    }
    const storage = typeof arg === "string" ? new HttpStorage(arg) : arg;
    const requestOptions = options.signal ? { signal: options.signal } : void 0;
    const repo = await Repository.open(
      { storage, formatVersion: options.formatVersion },
      requestOptions
    );
    let session;
    if (options.snapshot) {
      session = await repo.checkoutSnapshot(options.snapshot, requestOptions);
    } else if (options.tag) {
      session = await repo.checkoutTag(options.tag, requestOptions);
    } else {
      session = await repo.checkoutBranch(
        options.branch ?? "main",
        requestOptions
      );
    }
    return new _IcechunkStore(session, options.transformRequest);
  }
  /**
   * Get data for a zarr key.
   *
   * zarrita calls this method with keys like:
   * - "/zarr.json" - root metadata
   * - "/array/zarr.json" - array metadata
   * - "/array/c/0/1/2" - chunk at coords [0, 1, 2]
   *
   * @param key - Absolute path (must start with "/")
   * @param opts - Optional request options (supports AbortSignal for cancellation)
   * @returns Data bytes or undefined if not found
   */
  async get(key, opts) {
    if (opts?.signal?.aborted) return void 0;
    const parsed = parseZarrKey(key);
    try {
      if (parsed.type === "metadata") {
        const data = this.session.getRawMetadata(parsed.path);
        return data ?? void 0;
      }
      const chunk = await this.session.getChunk(parsed.path, parsed.coords, {
        signal: opts?.signal,
        transformRequest: this.transformRequest
      });
      return chunk ?? void 0;
    } catch {
      return void 0;
    }
  }
  /**
   * Get partial data for a zarr key.
   *
   * Required by zarrita for sharded arrays. zarrita uses this to read the
   * shard index (via suffixLength) and extract individual inner chunks
   * (via offset/length) from shard data.
   *
   * @param key - Absolute path
   * @param range - Byte range to fetch
   * @param opts - Optional request options (supports AbortSignal for cancellation)
   * @returns Data bytes or undefined if not found
   */
  async getRange(key, range, opts) {
    if (opts?.signal?.aborted) return void 0;
    const parsed = parseZarrKey(key);
    try {
      if (parsed.type === "chunk") {
        const data2 = await this.session.getChunkRange(
          parsed.path,
          parsed.coords,
          range,
          {
            signal: opts?.signal,
            transformRequest: this.transformRequest
          }
        );
        return data2 ?? void 0;
      }
      const data = this.session.getRawMetadata(parsed.path);
      if (!data) return void 0;
      if ("suffixLength" in range) {
        return data.slice(-range.suffixLength);
      }
      return data.slice(range.offset, range.offset + range.length);
    } catch {
      return void 0;
    }
  }
};
function parseZarrKey(key) {
  const path = key.slice(1);
  if (path === "zarr.json") {
    return { type: "metadata", path: "/" };
  }
  if (path.endsWith("/zarr.json")) {
    const nodePath = "/" + path.slice(0, -"/zarr.json".length);
    return { type: "metadata", path: nodePath };
  }
  const chunkMatch = path.match(/^(?:(.*)\/)?c(?:\/(.*))?$/);
  if (chunkMatch) {
    const arrayPath = chunkMatch[1] ? "/" + chunkMatch[1] : "/";
    const coordsStr = chunkMatch[2] ?? "";
    const coords = coordsStr ? coordsStr.split("/").map(Number) : [];
    return { type: "chunk", path: arrayPath, coords };
  }
  return { type: "metadata", path: "/" + path };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CompressionAlgorithm,
  FileType,
  HeaderParseError,
  HttpStorage,
  IcechunkStore,
  NotFoundError,
  ReadSession,
  Repository,
  SpecVersion,
  StorageError,
  decodeBase32,
  decodeObjectId12,
  decodeObjectId8,
  encodeBase32,
  encodeObjectId12,
  encodeObjectId8
});
